[BugFix] Introduce tablet_schema into MorselQueue for tablet_internal_parallel in clould native table

srlch opened the pull request:
## Why I'm doing:
In current impl tablet_internal_parallel in clould native table, LogicalSplitMorselQueue/PhysicalSplitMorselQueue
will use lake::Tablet::tablet_schema to get tablet_schema info to parse the seek range. But the problem is that,
lake::Tablet::tablet_schema will return nullptr in some case and make BE/CN crash when parse the range.

Beside, lake::Tablet::tablet_schema can not ensure the version of tablet schema is same as the version in lake::TabletReader.

## What I'm doing:
Introduce tablet_schema into MorselQueue by hand instead using (lake::)Tablet::tablet_schema to get tablet_schema.

Fixes #issue

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.4
  - [x] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0

trueeyu commented on be/src/exec/pipeline/scan/morsel.h:
```diff
@@ -360,6 +360,10 @@ class MorselQueue {
     virtual Status append_morsels(Morsels&& morsels);
     virtual Type type() const = 0;
+    void set_tablet_schema(TabletSchemaCSPtr tablet_schema) {
+        DCHECK(tablet_schema.get() != nullptr);
```
DCHECK(tablet_schema != nullptr);

trueeyu commented on be/src/exec/pipeline/scan/morsel.cpp:
```diff
@@ -851,7 +863,7 @@ Status LogicalSplitMorselQueue::_init_tablet() {
 
     _short_key_schema =
-            std::make_shared<Schema>(ChunkHelper::get_short_key_schema(_tablets[_tablet_idx]->tablet_schema()));
+            std::make_shared<Schema>(ChunkHelper::get_short_key_schema(_tablet_schema));
```
In what scenarios, the return value of `_tablets[_tablet_idx]->tablet_schema()` is nullptr?

srlch replied:
cache miss and then network failure, or tablet has been dropped by catalogBin

trueeyu commented on be/src/exec/pipeline/scan/morsel.cpp:
```diff
@@ -504,9 +504,15 @@ Status PhysicalSplitMorselQueue::_init_segment() {
+            if (!_tablets[_tablet_idx]->belonged_to_cloud_native()) {
+                RETURN_IF_ERROR(TabletReader::parse_seek_range(_tablet_schema, _range_start_op, _range_end_op,
+                                                               _range_start_key, _range_end_key, &_tablet_seek_ranges,
+                                                               &_mempool));
```
If fixed two bugs, better to split to 2 prs.

srlch replied:
ok

srlch committed:
[BugFix] Introduce tablet_schema into MorselQueue for tablet_internal_parallel in clould native table

## Why I'm doing:
In current impl tablet_internal_parallel in clould native table, LogicalSplitMorselQueue/PhysicalSplitMorselQueue
will use lake::Tablet::tablet_schema to get tablet_schema info to parse the seek range. But the problem is that,
lake::Tablet::tablet_schema will return nullptr in some case and make BE/CN crash when parse the range.

Beside, lake::Tablet::tablet_schema can not ensure the version of tablet schema is same as the version in lake::TabletReader.

## What I'm doing:
Introduce tablet_schema into MorselQueue by hand instead using (lake::)Tablet::tablet_schema to get tablet_schema.

Signed-off-by: srlch <linzichao@starrocks.com>

srlch committed:
fix

Signed-off-by: srlch <linzichao@starrocks.com>

srlch committed:
fix

Signed-off-by: srlch <linzichao@starrocks.com>

srlch committed:
fix

Signed-off-by: srlch <linzichao@starrocks.com>
